### PR TITLE
Make it possible to mark all beans based on test classes as unremovable

### LINKED ISSUE #1648: Error injecting beans in test sources
Having a bean defined in test sources that depends on the EM:

```java
@ApplicationScoped
public class TestDataFactory {

    @Inject
    protected EntityManager em;
}
```

and trying to inject it into a @QuarkusTest I got this Exception:

[io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [net.lucamasini.testcase.QuarkusResourceTest]: Failed to inject field protected net.lucamasini.testcase.TestDataFactory net.lucamasini.testcase.QuarkusResourceTest.dataFactory
Caused by: java.lang.RuntimeException: Failed to inject field protected net.lucamasini.testcase.TestDataFactory net.lucamasini.testcase.QuarkusResourceTest.dataFactory
Caused by: java.lang.NullPointerException: Managed Bean [class net.lucamasini.testcase.TestDataFactory] is null

Attached a test case.
[test-case.zip](https://github.com/quarkusio/quarkus/files/2998519/test-case.zip)

### PR DESCRIPTION
**Pull request by mkouba**
- resolves https://github.com/quarkusio/quarkus/issues/1648

### CODE REVIEW

File: test-framework/common/src/main/java/io/quarkus/test/common/PathTestHelper.java [hunk 1]
```diff
@@ -63,4 +63,23 @@ public static Path getAppClassLocation(Class<?> testClass) {
...
+
+    public static boolean isTestClass(String className, ClassLoader classLoader) {
+        String classFileName = className.replace('.', File.separatorChar) + ".class";
+        URL resource = classLoader.getResource(classFileName);
+        if (!resource.getProtocol().startsWith("file")) {
+            return false;
+        }
+        try {
+            Path path = Paths.get(resource.toURI());
+            if (path.toString().contains(TEST_CLASSES_FRAGMENT_MAVEN) ||
```
Comment (geoand) [inline test-framework/common/src/main/java/io/quarkus/test/common/PathTestHelper.java]: Perhaps simplify the if statement to: 
```
return path.toString().contains(TEST_CLASSES_FRAGMENT_MAVEN) ||
                    path.toString().contains(TEST_CLASSES_FRAGMENT_GRADLE);
```

?

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcTestResourceProvider.java [hunk 2]
```diff
@@ -4,38 +4,34 @@
...
+
     @Override
     public void inject(Object test) {
         Class<?> c = test.getClass();
+        ArcContainer container = Arc.container();
         while (c != Object.class) {
             for (Field f : c.getDeclaredFields()) {
                 if (f.isAnnotationPresent(Inject.class)) {
                     try {
-                        BeanManager beanManager = Arc.container().beanManager();
```
Comment (geoand) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcTestResourceProvider.java]: :+1: :+1: 

Makes this much cleaner IMHO. When looking at various pieces of the code in the past I never really understood when to use Arc and when to use the BeanManager. This makes things a lot cleaner I think!

**Comment by geoand** [comment]
@mkouba Assuming the tests will be fixed (which I am guessing by the failure should be pretty easy), the handling of test resources seems great, :+1: !

I can't really comment on the `RequestContextController` stuff since I don't even know what it is :)

**Comment by mkouba** [comment]
@manovotn Pls review the `RequestContextController` part.

**Review by geoand — APPROVED** [review]

**Review by manovotn — COMMENTED** [review]
@mkouba looks good!
I suppose Arc doesn't do `Initialized` and `Destroyed` events for contexts, right?

**Comment by mkouba** [comment]
@manovotn We do fire these events but only for `ApplicationScoped`. It should be quite straightforward to implement this for other scopes:
https://github.com/quarkusio/quarkus/issues/1680

:-)

**Comment by manovotn** [comment]
Will do! ;-)

**Comment by geoand** [comment]
So I guess we should merge this now?

**Pull request merged**